karapace/protobuf/io.py

```diff
@@ -0,0 +1,206 @@
+                return 0
+            raise EOFError(f"EOF while reading varint, value is {varint} so far")
+
+        byte = ord(char)
+        varint += (byte & 0x7F) << (7 * read_bytes)
+
+        read_bytes += 1
+
+        if not byte & 0x80:
+            return varint
+
+
+def read_indexes(bio: BytesIO):
+    try:
+        size: int = read_varint(bio)
+    except EOFError:
+        # TODO: change exception
+        raise IllegalArgumentException("problem with reading binary data")
+    if size == 0:
+        return [0]
+    return [read_varint(bio) for _ in range(size)]
+
+
+def find_message_name(schema: ProtobufSchema, indexes: List[int]) -> str:
+    result: list = []
+    types = schema.proto_file_element.types
+    for index in indexes:
+        try:
+            message = types[index]
+        except IndexError:
+            raise IllegalArgumentException(f"Invalid message indexes: {indexes}")
+
+        if message and isinstance(message, MessageElement):
+            result.append(message.name)
+            types = message.nested_types
+        else:
+            raise IllegalArgumentException(f"Invalid message indexes: {indexes}")
+
+    # for java we also need package name. But in case we will use protoc
+    # for compiling to python we can ignore it at all
+    return ".".join(result)
+
+
+def get_protobuf_class_instance(schema: ProtobufSchema, class_name: str, cfg: Dict) -> Any:
+    directory = cfg["protobuf_runtime_directory"]
+    proto_name = calculate_class_name(str(schema))
+    proto_path = f"{directory}/{proto_name}.proto"
+    class_path = f"{directory}/{proto_name}_pb2.py"
+    if not os.path.isfile(proto_path):
+        with open(f"{directory}/{proto_name}.proto", "w") as proto_text:
+            proto_text.write(str(schema))
+
+    if not os.path.isfile(class_path):
+        subprocess.run([
+            "protoc",
+            "--python_out=./",
+            proto_path,
+        ], check=True)
+
+    spec = importlib.util.spec_from_file_location(f"{proto_name}_pb2", class_path)
+    tmp_module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(tmp_module)
+    class_to_call = getattr(tmp_module, class_name)
+    return class_to_call()
+
+
+def read_data(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema, bio: BytesIO):
+    # TODO (serge): check and polish it
+    if not match_schemas(writer_schema, reader_schema):
+        fail_msg = 'Schemas do not match.'
+        raise ProtobufSchemaResolutionException(fail_msg, writer_schema, reader_schema)
+
+    indexes = read_indexes(bio)
+    name = find_message_name(writer_schema, indexes)
+
+    class_instance = get_protobuf_class_instance(writer_schema, name, config.DEFAULTS)
+    class_instance.ParseFromString(bio.read())
+
+    return class_instance
+
+
+class ProtobufDatumReader:
+    """Deserialize Protobuf-encoded data into a Python data structure."""
+
+    def __init__(self, writer_schema=None, reader_schema=None):
```

Comment:
Please make both args required, their types are `ProtobufSchema`:

```suggestion
    def __init__(self, writer_schema: ProtobufSchema, reader_schema: ProtobufSchema) -> None:
```